File: step_13_binary_trees/13.3_hard/P010_build_bt_from_inorder_postorder/solution.py
"""
Problem: Construct Binary Tree from Inorder and Postorder Traversal
Difficulty: MEDIUM | XP: 25
LeetCode: #106

Given two integer arrays inorder and postorder (both of the same binary tree),
construct and return the binary tree.

Key insight:
- Postorder: [left][right][ROOT] -- last element is always the root
- Inorder:   [left][ROOT][right] -- root splits it into left and right subtrees
"""
from typing import List, Optional
from collections import deque


class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right


# ============================================================
# APPROACH 1: BRUTE FORCE -- Recursive with linear inorder search
# Time: O(n^2)  |  Space: O(n)
# ============================================================
def build_tree_brute(inorder: List[int], postorder: List[int]) -> Optional[TreeNode]:
    """
    Recursively:
    1. Last element of postorder is the root.
    2. Find it in inorder with a linear scan -> splits into left/right.
    3. Recurse on left and right sublists.
    """
    if not inorder or not postorder:
        return None

    root_val = postorder[-1]
    root = TreeNode(root_val)

    # Find root in inorder -- O(n) scan
    mid = inorder.index(root_val)

    # Elements left of mid in inorder = left subtree
    # Corresponding postorder chunk has the same count
    left_size = mid
    root.left = build_tree_brute(inorder[:mid], postorder[:left_size])
    root.right = build_tree_brute(inorder[mid + 1:], postorder[left_size:-1])

    return root


# ============================================================
# APPROACH 2: OPTIMAL -- HashMap for O(1) inorder lookup
# Time: O(n)  |  Space: O(n)
# ============================================================
def build_tree_optimal(inorder: List[int], postorder: List[int]) -> Optional[TreeNode]:
    """
    Precompute a map of value -> index in inorder for O(1) lookups.
    Pass index boundaries instead of slicing arrays.
    """
    idx_map = {val: i for i, val in enumerate(inorder)}
    post_idx = [len(postorder) - 1]  # mutable pointer into postorder (rightmost first)

    def build(in_lo: int, in_hi: int) -> Optional[TreeNode]:
        if in_lo > in_hi:
            return None

        root_val = postorder[post_idx[0]]
        post_idx[0] -= 1
        root = TreeNode(root_val)

        mid = idx_map[root_val]

        # IMPORTANT: build right subtree FIRST because postorder is processed right-to-left
        root.right = build(mid + 1, in_hi)
        root.left = build(in_lo, mid - 1)

        return root

    return build(0, len(inorder) - 1)


# ============================================================
# APPROACH 3: BEST -- Same as Optimal, explicit post_end parameter
# Time: O(n)  |  Space: O(n) -- cleaner without mutable list trick
# ============================================================
def build_tree_best(inorder: List[int], postorder: List[int]) -> Optional[TreeNode]:
    """
    Pass postorder end index explicitly through the recursion.
    Same O(n) time/space but slightly cleaner parameterization.
    The postorder end index shrinks by (left_size + right_size + 1) at each level.
    """
    idx_map = {val: i for i, val in enumerate(inorder)}

    def build(in_lo: int, in_hi: int, post_end: int) -> Optional[TreeNode]:
        if in_lo > in_hi:
            return None

        root_val = postorder[post_end]
        root = TreeNode(root_val)
        mid = idx_map[root_val]

        left_size = mid - in_lo
        right_size = in_hi - mid

        # Right subtree uses postorder[post_end - right_size .. post_end - 1]
        root.right = build(mid + 1, in_hi, post_end - 1)
        # Left subtree uses postorder[post_end - right_size - left_size .. post_end - right_size - 1]
        root.left = build(in_lo, mid - 1, post_end - right_size - 1)

        return root

    if not inorder:
        return None
    return build(0, len(inorder) - 1, len(postorder) - 1)


# ============================================================
# Helpers
# ============================================================
def inorder_of(root: Optional[TreeNode]) -> List[int]:
    res = []
    def dfs(node):
        if node:
            dfs(node.left)
            res.append(node.val)
            dfs(node.right)
    dfs(root)
    return res


def postorder_of(root: Optional[TreeNode]) -> List[int]:
    res = []
    def dfs(node):
        if node:
            dfs(node.left)
            dfs(node.right)
            res.append(node.val)
    dfs(root)
    return res


if __name__ == "__main__":
    print("=== Build Binary Tree from Inorder and Postorder ===\n")

    tests = [
        ([9, 3, 15, 20, 7], [9, 15, 7, 20, 3]),   # LC #106 example
        ([2, 1, 3], [2, 3, 1]),                     # root=1
        ([1], [1]),                                 # single node
        ([-1], [-1]),                               # negative
    ]

    for inorder, postorder in tests:
        r1 = build_tree_brute(inorder[:], postorder[:])
        r2 = build_tree_optimal(inorder[:], postorder[:])
        r3 = build_tree_best(inorder[:], postorder[:])

        io1, po1 = inorder_of(r1), postorder_of(r1)
        io2, po2 = inorder_of(r2), postorder_of(r2)
        io3, po3 = inorder_of(r3), postorder_of(r3)

        ok = (io1 == io2 == io3 == inorder) and (po1 == po2 == po3 == postorder)
        print(f"inorder={inorder}, postorder={postorder}")
        print(f"  Brute   -> inorder={io1}, postorder={po1}")
        print(f"  Optimal -> inorder={io2}, postorder={po2}")
        print(f"  Best    -> inorder={io3}, postorder={po3}")
        print(f"  {'PASS' if ok else 'FAIL'}\n")

File: step_13_binary_trees/13.3_hard/P010_build_bt_from_inorder_postorder/Solution.java
/**
 * Problem: Construct Binary Tree from Inorder and Postorder Traversal
 * Difficulty: MEDIUM | XP: 25
 * LeetCode: #106
 *
 * Given inorder and postorder traversal arrays of a binary tree,
 * construct and return the binary tree.
 *
 * Key insight:
 *  - Postorder: [left][right][ROOT] -- last element is always the root
 *  - Inorder:   [left][ROOT][right] -- root splits into left/right subtrees
 *
 * @author DSA_Nova
 */
import java.util.*;

public class Solution {

    static class TreeNode {
        int val;
        TreeNode left, right;
        TreeNode(int val) { this.val = val; }
    }

    // ============================================================
    // APPROACH 1: BRUTE FORCE -- Recursive with linear inorder search
    // Time: O(n^2)  |  Space: O(n)
    // ============================================================
    public static TreeNode buildTreeBrute(int[] inorder, int[] postorder) {
        return buildBrute(inorder, 0, inorder.length - 1,
                          postorder, 0, postorder.length - 1);
    }

    private static TreeNode buildBrute(int[] inorder, int inLo, int inHi,
                                        int[] postorder, int postLo, int postHi) {
        if (inLo > inHi || postLo > postHi) return null;

        int rootVal = postorder[postHi];
        TreeNode root = new TreeNode(rootVal);

        // Linear scan in inorder for root
        int mid = inLo;
        while (mid <= inHi && inorder[mid] != rootVal) mid++;

        int leftSize = mid - inLo;

        root.left  = buildBrute(inorder, inLo, mid - 1,
                                postorder, postLo, postLo + leftSize - 1);
        root.right = buildBrute(inorder, mid + 1, inHi,
                                postorder, postLo + leftSize, postHi - 1);
        return root;
    }

    // ============================================================
    // APPROACH 2: OPTIMAL -- HashMap for O(1) inorder lookup
    // Time: O(n)  |  Space: O(n)
    // ============================================================
    static Map<Integer, Integer> idxMap;
    static int[] postArr;
    static int postIdx;

    public static TreeNode buildTreeOptimal(int[] inorder, int[] postorder) {
        idxMap = new HashMap<>();
        for (int i = 0; i < inorder.length; i++) idxMap.put(inorder[i], i);
        postArr = postorder;
        postIdx = postorder.length - 1;
        return buildOptimal(0, inorder.length - 1);
    }

    private static TreeNode buildOptimal(int inLo, int inHi) {
        if (inLo > inHi) return null;

        int rootVal = postArr[postIdx--];
        TreeNode root = new TreeNode(rootVal);
        int mid = idxMap.get(rootVal);

        // Build right FIRST (postorder is consumed right-to-left)
        root.right = buildOptimal(mid + 1, inHi);
        root.left  = buildOptimal(inLo, mid - 1);
        return root;
    }

    // ============================================================
    // APPROACH 3: BEST -- HashMap + explicit postEnd parameter (no global state)
    // Time: O(n)  |  Space: O(n)
    // ============================================================
    public static TreeNode buildTreeBest(int[] inorder, int[] postorder) {
        if (inorder.length == 0) return null;
        Map<Integer, Integer> map = new HashMap<>();
        for (int i = 0; i < inorder.length; i++) map.put(inorder[i], i);
        return buildBest(inorder, postorder, map, 0, inorder.length - 1, postorder.length - 1);
    }

    private static TreeNode buildBest(int[] inorder, int[] postorder,
                                       Map<Integer, Integer> map,
                                       int inLo, int inHi, int postEnd) {
        if (inLo > inHi) return null;

        int rootVal = postorder[postEnd];
        TreeNode root = new TreeNode(rootVal);
        int mid = map.get(rootVal);
        int rightSize = inHi - mid;

        root.right = buildBest(inorder, postorder, map, mid + 1, inHi, postEnd - 1);
        root.left  = buildBest(inorder, postorder, map, inLo, mid - 1, postEnd - rightSize - 1);
        return root;
    }

    // ============================================================
    // Helpers
    // ============================================================
    public static List<Integer> inorderOf(TreeNode root) {
        List<Integer> res = new ArrayList<>();
        inHelper(root, res);
        return res;
    }

    private static void inHelper(TreeNode node, List<Integer> res) {
        if (node == null) return;
        inHelper(node.left, res);
        res.add(node.val);
        inHelper(node.right, res);
    }

    public static List<Integer> postorderOf(TreeNode root) {
        List<Integer> res = new ArrayList<>();
        postHelper(root, res);
        return res;
    }

    private static void postHelper(TreeNode node, List<Integer> res) {
        if (node == null) return;
        postHelper(node.left, res);
        postHelper(node.right, res);
        res.add(node.val);
    }

    public static void main(String[] args) {
        System.out.println("=== Build Binary Tree from Inorder and Postorder ===\n");

        int[][] inorders   = {{9,3,15,20,7}, {2,1,3}, {1}, {-1}};
        int[][] postorders = {{9,15,7,20,3}, {2,3,1}, {1}, {-1}};

        for (int t = 0; t < inorders.length; t++) {
            int[] inorder = inorders[t];
            int[] postorder = postorders[t];

            TreeNode r1 = buildTreeBrute(inorder, postorder);
            TreeNode r2 = buildTreeOptimal(inorder, postorder);
            TreeNode r3 = buildTreeBest(inorder, postorder);

            List<Integer> io1 = inorderOf(r1),  po1 = postorderOf(r1);
            List<Integer> io2 = inorderOf(r2),  po2 = postorderOf(r2);
            List<Integer> io3 = inorderOf(r3),  po3 = postorderOf(r3);

            // Convert expected arrays to lists
            List<Integer> expIn = new ArrayList<>(), expPost = new ArrayList<>();
            for (int v : inorder)   expIn.add(v);
            for (int v : postorder) expPost.add(v);

            boolean ok = io1.equals(expIn) && io2.equals(expIn) && io3.equals(expIn)
                      && po1.equals(expPost) && po2.equals(expPost) && po3.equals(expPost);
            System.out.printf("inorder=%s postorder=%s%n",
                Arrays.toString(inorder), Arrays.toString(postorder));
            System.out.printf("  Brute   inorder=%s postorder=%s%n", io1, po1);
            System.out.printf("  Optimal inorder=%s postorder=%s%n", io2, po2);
            System.out.printf("  Best    inorder=%s postorder=%s%n", io3, po3);
            System.out.printf("  %s%n%n", ok ? "PASS" : "FAIL");
        }
    }
}
